rustls/src/quic.rs

```diff
@@ -108,8 +164,66 @@ impl PacketKey {
                 &[],
             )),
             iv: hkdf_expand(secret, IvLen, b"quic iv", &[]),
+            suite,
         }
     }
+
+    /// Encrypt a QUIC packet
+    ///
+    /// `buf` contains the header, payload and space for the AEAD tag, with the payload starting at
+    /// `header_len` and ending at `tag_len()` before the end of `buf`.
+    pub fn encrypt(&self, packet: u64, buf: &mut [u8], header_len: usize) -> Result<(), Error> {
```

Comment:
This API seems really error prone as designed as it requires the caller to do a bunch of slicing.

It also seems potentially suboptimal because it precludes the use of IO vectors for the header and tag to minimize the need to copy data.

AFAICT, it would be better to simply have `header: &[u8]` as a separate parameter; the caller can do `split_at_mut` themselves if they want the header and the rest to be in a single buffer.

Similarly, I think if we just returned the tag (in a newtype around `ring::aead::Tag`) then the caller could decide whether to use IOVecs or to append to the buffer.

This would eliminate all the slicing and dicing and potential for things to go wrong in the caller's invocation of `encrypt`.

I think having "_in_place" in the name would also make it clearer what's happening.

Reply:
Fixed up as suggested.